Mathlib/Data/Polynomial/Derivation.lean

```diff
@@ -92,4 +92,106 @@ def mkDerivationEquiv : A ≃ₗ[R] Derivation R R[X] A :=
 @[simp] lemma mkDerivationEquiv_symm_apply (D : Derivation R R[X] A) :
     (mkDerivationEquiv R).symm D = D X := rfl
 
+end CommSemiring
 end Polynomial
+
+namespace Module
+
+open Polynomial
+/--
+Suppose `a` is an element of an `R`-algebra `A` and `M` is an `A`-module.
+Then `Module.comp_aeval R M a` is the `R[X]`-module with carrier `M`,
+where the action of `f : R[X]` is `f • m = (aeval a f) • m`.
+-/
+structure CompAEval (R M: Type*) {A : Type*} [CommSemiring R] [Semiring A] [Algebra R A]
```

Comment:
This is an important construction that should belong to a more basic file. Moreover, are you sure this isn't already in mathlib in some form? I am not sure, but we should have something like the `R[X]`-module structure given by an endomorphism of `M` (or at least there has been a discussion about it).

Reply:
Ah, I just found [this](https://github.com/leanprover-community/mathlib/pull/14666) mathlib3 PR that for some reason never got merged. It's probably worth to try to do everything in the right generality.

Reply:
Thank you very much for your comments; it's taken me a long time to think them through. I have moved the module construction to Data.Polynomial.Module

As you say, my construction of an `R[X]`-module structure could be regarded as a special case of
a more general construction, starting from a linear map (as in the PR that you found for me).

Alternatively, one could view it the other way around: if one begins with a linear map φ,
then one could construct the corresponding `R[X]`-module as `CompAEval R M φ`. In fact, when I looked at the mathlib3 PR which you found for me (and many thanks for that!), I can see on line 60 that the `R[X]`-module structure is defined by composing with aeval in the case that the `R`-algebra is `M →ₗ[R] M`.

So each construction could be regarded as a generalization of the other. If you add the following line following line to the end of the file, then this will be clear:

`example (φ : M →ₗ[R] M) (m : CompAEval R M φ) : (X : R[X]) • m = φ • m := by simp`

On line 409 of Data.Polynomial.Module I have now added a definition of the `R[X]`-module corresponding to a linear map `φ : M →ₗ[R] M`.